src/main/java/it/smartcommunitylab/aac/saml/provider/SamlAuthenticationProvider.java

```diff
@@ -203,6 +236,32 @@ public Authentication doAuthenticate(Authentication authentication) throws Authe
         }
     }
 
+    private static List<Object> extractAcrAssertion(Response response) {
+        List<Object> acrValues = new ArrayList<>();
+        if ((response == null) || (response.getAssertions() == null) || (response.getAssertions().isEmpty())) {
+            return acrValues;
+        }
+        for (Assertion assertion : response.getAssertions()) {
+            for (Statement statement : assertion.getStatements()) {
+                if (!(statement instanceof AuthnStatementImpl)) {
+                    continue;
+                }
+                AuthnContext authnContext = ((AuthnStatementImpl) statement).getAuthnContext();
+                if (authnContext == null) {
+                    continue;
+                }
+                AuthnContextClassRef acr = authnContext.getAuthnContextClassRef();
+                if (acr == null) {
+                    continue;
+                }
+                if (StringUtils.hasText(acr.getURI())) {
+                    acrValues.add(acr.getURI());
+                }
+            }
+        }
+        return acrValues;
```

Comment:
do not add acr values from different statements, the result will be inconsistent. You return a list from the first non null statement, and skip others, with *exactly* the same content (as URI:string) read from the statement.

Reply:
Thanks for the suggestion.
Small note: one AuthnStatement can have at most one AuthnContext, hence one AuthnContextClassRef, hence one String ACR URI value.
Since taking "the list" induced by the **first** non nullable statement will always result in a list with up to **one** String ACR URI element, maybe it should be appropriate to change the function signature to return a nullable String (and update the caller as a consequence).